projects/transformers/trainer_mixins/distillation.py

```diff
@@ -0,0 +1,431 @@
+        output = torch.autograd.Variable(out, requires_grad=True)
+
+        target = torch.FloatTensor([[0.05, 0.9, 0.05], [0.05, 0.05, 0.9]])
+        target = torch.autograd.Variable(y1)
+        loss = cross_entropy(output, target)
+        loss.backward()
+
+    see: https://discuss.pytorch.org/t/cross-entropy-with-one-hot-targets/13580/5
+    """
+    if reduction == "mean":
+        return torch.mean(torch.sum(-target * F.log_softmax(output, dim=1), dim=1))
+    elif reduction == "sum":
+        return torch.sum(torch.sum(-target * F.log_softmax(output, dim=1), dim=1))
+    else:
+        raise ValueError(f"Unknown reduction: {reduction}")
+
+
+def masked_soft_cross_entropy(output, target, padding_mask):
+    """
+    Soft cross entropy which accepts a padding mask. Loss is scaled to take into
+    account the actual number of elements used to calculate the loss.
+
+    Args:
+    :param output: predictions for neural network
+    :param targets: targets, can be soft
+    :param padding_mask: mask used to pad loss functions for tokens which are not
+                         part of the label for that batch
+    """
+    loss_vec = -target * F.log_softmax(output, dim=-1)
+    loss_vec.masked_fill_(padding_mask, 0.0)
+    loss = loss_vec.sum()
+    num_active_elements = padding_mask.numel() - padding_mask.sum()
+    return loss / num_active_elements
+
+
+def calculate_linear_decay(
+    first_step_value, last_step_value, current_step, total_steps
+):
+    """
+    Calculates value for a current step in a linear decay.
+
+    :param first_step_value: Value at first step (before training).
+    :param last_step_value: Value at last step (after training).
+    :param current_step: Current step. Assumes first step is 0.
+    :param total_steps: Total number of steps in training.
+    """
+    step_size = (first_step_value - last_step_value) / (total_steps - 1)
+    return first_step_value - step_size * current_step
+
+
+def resize_position_embeddings(model, new_seq_length):
+    """
+    Resizes model's position embeddings matrices if the size of max position embedding
+    doesn't match new sequence length.
+    (size of position embedding equals size of the attention window)
+
+    :param :new_seq_length Tokenizer sequence length.
+    """
+
+    # Functions to recursively get and set attributes. Source:
+    # https://stackoverflow.com/questions/31174295/getattr-and-setattr-on-nested-subobjects-chained-properties  # noqa: E501
+    def rsetattr(obj, attr, val):
+        pre, _, post = attr.rpartition(".")
+        return setattr(rgetattr(obj, pre) if pre else obj, post, val)
+
+    def rgetattr(obj, attr, *args):
+        def _getattr(obj, attr):
+            return getattr(obj, attr, *args)
+        return functools.reduce(_getattr, [obj] + attr.split("."))
+
+    # Find and replace all position embeddings if the size of position embedding
+    # doesn't match new sequence length
+    for module_name, module in model.named_modules():
```

Comment:
That is good to know, thanks. They seem to do the same thing. You did showed me this function before, but I didn't make the connection while coding that this would be the perfect use case.